packages/migrations/README.md

```diff
@@ -0,0 +1,130 @@
+- Schema migrations are expected to run synchronously during deploy. So even if you wrote JavaScript code to manually batch up a table to avoid locks, you'd have to babysit a long-running process.
+- If errors are encountered, you'll have to figure out how to manually retry the change for the failing batches.
+
+By using batched migrations, these problems are avoided:
+
+- Work is done in small batches, so large numbers of rows (or even entire tables) are not locked for long periods of time.
+- Work is done asynchronously in the background, so migrations that operate on very large tables won't block deploys.
+- Each batch is tracked independently and failing batched can be easily retried.
+
+#### Writing batched migrations
+
+Batched migrations are written as an object with two functions:
+
+- `getParameters()`: returns the minimum and maximum IDs to operate on, as well as a batch size. `min` defaults to `1` and `batchSize` defaults to `1000`. If `max === null`, that indicates that there are no rows to operate on.
+- `execute(min, max)`: runs the migration on the given range of IDs, inclusive of its endpoints.
+
+A `makeBatchedMigration()` function is available to help ensure you're writing an object with the correct shape.
+
+```ts
+// batched-migrations/20230411002409_example_migration.ts
+import { makeBatchedMigration } from '@prairielearn/migrations';
+import { queryOneRowAsync, queryAsync } from '@prairielearn/postgres';
+
+export default makeBatchedMigration({
+  async getParameters() {
+    const result = await queryOneRowAsync('SELECT MAX(id) as max from examples;', {});
+    return {
+      max: result.rows[0].max,
+      batchSize: 1000,
+    };
+  },
+
+  async execute(min: bigint, max: bigint) {
+    await queryAsync('UPDATE examples SET text = TRIM(text) WHERE id >= $min AND id <= $max', {
+      min,
+      max,
+    });
+  },
+});
+```
+
+Batched migration `execute()` functions **must** be idempotent, as they may run multiple times on the same ID range in the case of retries after failure.
+
+#### Executing batched migrations
+
+You'll need to initialize and start batched migration processing in your application startup code. Since batched migrations are typically used with regular migrations, you'll need to take care to call `init()` after `initBatchedMigrations()` but before `startBatchedMigrations()`.
+
+```ts
+import { init, initBatchedMigrations, startBatchedMigrations } from '@prairielearn/migrations';
+
+const runner = initBatchedMigrations({
+  project: 'prairielearn',
+  directories: [path.join(__dirname, 'batched-migrations')],
+});
+runner.on('error', (error) => {
+  // Handle error, e.g. by reporting to Sentry.
+});
+
+await init();
+
+startBatchedMigrations({
+  workDurationMs: 60_000,
+  sleepDurationMs: 30_000,
+});
+```
```

Comment:
I'm not sure if this example is needed/valuable here. I guess it is a question of the difference between documentation for someone wanting to write a batched migration within PrairieLearn, and documentation for someone wanting to use this package in a different system.

As someone that wants to implement a batched migration within PrairieLearn, having this here confused me a little bit because I was wondering if I should be doing this, only to realize I didn't need to do this part because it is already done by PrairieLearn. 

Maybe instead of having the code example here, you could just end the previous paragraph with "for an example, see how this is implemented in PrairieLearn in `server.js`"

Reply:
I can move this to the end of the document if that would help surface the information that's going to be useful most often.

Reply:
How do the changes from 85c31ca look?

Reply:
Looks good! I still sort of think it would be nice to reference the implementation in "server.js" so people know where to look if they want to see it actually implemented but thats not critical.